js/utils.js

```diff
@@ -1511,47 +1511,58 @@ Fliplet.Registry.set('dynamicListUtils', (function() {
     });
   }
 
-  function setFilterValues(data) {
+  function setFilterValues(options) {
     return new Promise(function(resolve) {
-      if (!data.filterOptions.length) {
+      if (!(options.config.filterOptions || []).length) {
         return resolve();
       }
 
-      data.filterOptions.forEach(function(item) {
-        switch (item.valueType) {
-          case 'user-profile-data':
-            Fliplet.User.getCachedSession()
-              .then(function(options) {
-                var entries = options.entries;
-
-                if (options && options.entries) {
-                  if (entries.dataSource && entries.dataSource.data.hasOwnProperty(item.fieldValue)) {
-                    item.value = entries.dataSource.data[item.fieldValue];
-                    return resolve();
-                  }
-
-                  if (entries.saml2 && entries.saml2.data.hasOwnProperty(item.fieldValue)) {
-                    item.value = entries.saml2.data[item.fieldValue];
-                    return resolve();
-                  }
-
-                  if (entries.flipletLogin && entries.flipletLogin.data.hasOwnProperty(item.fieldValue)) {
-                    item.value = entries.flipletLogin.data[item.fieldValue];
-                    return resolve();
-                  }
-                }
-
-                Fliplet.Profile.get(item.fieldValue)
-                  .then(function(result) {
-                    item.value = result || '';
-                    return resolve();
+      var promises = [];
+
+      options.config.filterOptions.forEach(function(item, index) {
+        promises.push(
+          new Promise(function(res) {
+            switch (item.valueType) {
+              case 'user-profile-data':
+                Fliplet.User.getCachedSession()
```

Comment:
If there are 10 rules all using profile data, then `Fliplet.User.getCachedSession()` is being called 10 times. The result is the same so it really only need to be called once. On native apps calling this requires accessing native storage data and so 10 calls would create unnecessary lag.